### PR TITLE
Changed to MatrixBase from Matrix

### LINKED ISSUE #14696: error in calculating between an UGate operator and well-defined Unitary operators.
Hi,
I have a problem in calculating between an UGate operator and well-defined Unitary operators in  sympy-1.1.1-py3.6.

I tested two codes as the following:

```
>>> from sympy import *
>>> from sympy.physics.quantum import *
>>> from sympy.physics.quantum.qubit import Qubit
>>> from sympy.physics.quantum.gate import UGate,H
>>> u1 = UGate(0, ImmutableMatrix([[1/sqrt(2),1/sqrt(2)],[I/sqrt(2),-I/sqrt(2)]]))
>>> represent(H(0)*u1,nqubits=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/guest/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/sympy/physics/quantum/represent.py", line 207, in represent
    result = represent(expr.args[-1], **options)
  File "/Users/guest/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/sympy/physics/quantum/represent.py", line 145, in represent
    return expr._represent(**options)
  File "/Users/guest/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/sympy/physics/quantum/qexpr.py", line 393, in _represent
    result = self._format_represent(result, format)
  File "/Users/guest/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/sympy/physics/quantum/qexpr.py", line 398, in _format_represent
    return to_sympy(result)
  File "/Users/guest/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/sympy/physics/quantum/matrixutils.py", line 98, in to_sympy
    raise TypeError('Expected sympy/numpy/scipy.sparse matrix, got: %r' % m)
TypeError: Expected sympy/numpy/scipy.sparse matrix, got: None
```

or

```
>>> from sympy import *
>>> from sympy.physics.quantum import *
>>> from sympy.physics.quantum.gate import X,UGate
>>> v,w,f,g = symbols('v w f g')
>>> u2 = UGate(0,ImmutableMatrix([[v,w],[f,g]]))
>>> u3=X(0)+u2
>>> represent(u3,nqubits=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>  
  File "/Users/guest/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/sympy/physics/quantum/represent.py", line 165, in represent
    result = represent(expr.args[0], **options)
  File "/Users/guest/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/sympy/physics/quantum/represent.py", line 145, in represent
    return expr._represent(**options)
  File "/Users/guest/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/sympy/physics/quantum/qexpr.py", line 387, in _represent
    result = self._represent_default_basis(**options)
  File "/Users/guest/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/sympy/physics/quantum/gate.py", line 256, in _represent_default_basis
    return self._represent_ZGate(None, **options)
  File "/Users/guest/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/sympy/physics/quantum/gate.py", line 278, in _represent_ZGate
    controls, targets, target_matrix, nqubits, format
  File "/Users/guest/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/sympy/physics/quantum/gate.py", line 1094, in represent_zbasis
    return matrix_tensor_product(*product)
  File "/Users/guest/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/sympy/physics/quantum/matrixutils.py", line 247, in matrix_tensor_product
    return _sympy_tensor_product(*product)
  File "/Users/guest/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/sympy/physics/quantum/matrixutils.py", line 194, in _sympy_tensor_product
    'Sequence of Matrices expected, got: %s' % repr(matrices)
TypeError: Sequence of Matrices expected, got: (Matrix([
[1, 0],
[0, 1]]), Matrix([
[v, w],
[f, g]]))
```

According to some examination, I found an issue in physics/quantum/matrixutils.py.
Perhaps matrixutils.py had better to use DenseMatrix class instead of Matrix class. 
The test code would work well without any errors when I have replaced the 'Matrix' to 'DenseMatrix'.
I propose to fix physics/quantum/matrixutils.py.

Regards,

### PR DESCRIPTION
**Pull request by kyamaz**
#### References to other Issues or PRs

fixes #14696 

#### Brief description of what is fixed or changed
Replaced Matrix in physics.quantum.matrixutils by MatrixBase which is a common superclass of all SymPy matrices.

### CODE REVIEW

**Review by sidhantnagpal — COMMENTED** [review]
It would be good to include the test mentioned in the referenced issue, and additional tests if required.

**Comment by kyamaz** [comment]
I think I have fixed it and added test.

File: sympy/physics/quantum/tests/test_gate.py [hunk 1]
```diff
@@ -133,6 +133,19 @@ def test_UGate_CGate_combo():
...
             qubit_to_matrix(qapply(u2*IntQubit(i, 2)))
 
+def test_UGate_OneQubitGate_combo():
+    v,w,f,g = symbols('v w f g')
+    uMat1 = ImmutableMatrix([[v,w],[f,g]])
+    cMat1 = Matrix([[v, w + 1, 0, 0], [f + 1, g, 0, 0], [0, 0, v, w + 1], [0, 0, f + 1, g]])
+    u1 = X(0) + UGate(0, uMat1)
+    assert represent(u1, nqubits=2) == cMat1
+
+    uMat2 = ImmutableMatrix([[1/sqrt(2),1/sqrt(2)],[I/sqrt(2),-I/sqrt(2)]])
```
Comment (sidhantnagpal) [inline sympy/physics/quantum/tests/test_gate.py]: Do add spaces after commas throughout the function, otherwise looks fine.

**Comment by jksuom** [comment]
Thank you, this looks good.

**Pull request merged**